packages/ckeditor5-engine/src/model/document.ts

```diff
@@ -213,20 +247,29 @@ export default class Document extends EmitterMixin() {
 	/**
 	 * Returns a root by its name.
 	 *
-	 * @param name A unique root name.
+	 * Detached roots are returned by this method. This is to be able to operate on the detached root (for example, to be able to create
+	 * a position inside such a root for undo feature purposes).
+	 *
+	 * @param name The root name of the root to return.
 	 * @returns The root registered under a given name or `null` when there is no root with the given name.
 	 */
 	public getRoot( name: string = 'main' ): RootElement | null {
 		return this.roots.get( name );
 	}
 
 	/**
-	 * Returns an array with names of all roots (without the {@link #graveyard}) added to the document.
+	 * Returns an array with names of all roots added to the document (except the {@link #graveyard graveyard root}).
+	 *
+	 * Detached roots **are not** returned by this method by default. This is to make sure that all features or algorithms that operate
+	 * on the document data know which roots are still a part of the document and should be processed.
```

Comment:
Alternatively, we can refrain from changing this method and return all roots. Then, it will be less confusing vs `getRoot()` (which probably has to stay as it is) but it will be additional effort to make sure that all parts of the code that currently use `getRootNames()` will filter out detached roots if needed.

Another idea to avoid confusion would be to not return detached roots in `getRoot()`, however, then, some parts of the code (not sure how many) will have to use `document.roots.get()` directly. For sure it would have to happen when `Position` instance is created, as you may have positions inside detached roots.

I think what I proposed is the safest, although surely a bit confusing.